src/internet_identity/src/main.rs

```diff
@@ -368,6 +408,128 @@ async fn remove(user_number: UserNumber, device_key: DeviceKey) {
     })
 }
 
+// TODO: should this take a PoW to prevent spam?
+#[update]
+async fn create_challenge() -> Challenge {
+
+    let rng = make_rng().await;
+
+    let resp = STATE.with(|s| {
+        let mut inflight_challenges = s.inflight_challenges.borrow_mut();
+
+        // Prune old challenges. This drops all challenges that are older than
+        // CAPTCHA_CHALLENGE_LIFETIME
+        // TODO: test this
+        let now = time() as u64;
+        inflight_challenges.retain(|_, v| v.created > now - CAPTCHA_CHALLENGE_LIFETIME);
+
+        // Error out if there are too many inflight challenges
+        // TODO: test this
+        if inflight_challenges.len() >= MAX_INFLIGHT_CHALLENGES {
+            trap("too many inflight captchas");
+        }
+
+        // actually create the challenge
+
+        // first, try to find a key. We've already checked that we have fewer keys than
+        // MAX_INFLIGHT_CHALLENGES but just to make sure we use MAX_INFLIGHT_CHALLENGES as an upper
+        // bound of attempts.
+        let mut challenge_key: Option<u32> = None;
+        let mut attempt: u32 = 0;
+        while attempt < MAX_INFLIGHT_CHALLENGES as u32 {
+            if !inflight_challenges.contains_key(&attempt) {
+                challenge_key = Some(attempt);
+                break;
+            }
+            attempt += 1;
+        }
```

Comment:
We could write it like that, I just suck at rust :) We're not constraining the keys themselves, but we're constraining the number of max inflight challenges, so... instead of trying to find a free key forever we know the upper bound